birdie/src-tauri/src/repo/file.rs

```diff
@@ -42,10 +43,98 @@ pub struct File {
 
 impl PartialEq for File {
     fn eq(&self, other: &Self) -> bool {
-        self.name == other.name && self.size == other.size && self.file_type == other.file_type
+        self.path == other.path
+            && self.name == other.name
+            && self.size == other.size
+            && self.file_type == other.file_type
     }
 }
 
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SingleFileRequest {
+    pub path: String,
+}
+
+pub async fn get_file(
+    State(state): State<Arc<AppState>>,
+    Json(request): Json<SingleFileRequest>,
+) -> Result<Json<File>, CoreError> {
+    debug!("get_file: {:?}", request.path);
+    let repo_path = state.app_config.read().repo_path.clone();
+
+    let file_path = format!("{}/{}", repo_path, request.path.clone());
+
+    let file: File = {
+        let lock_cache = state.lock_cache.read().await;
+
+        let metadata = fs::metadata(file_path.clone())?;
+
+        let file_type = if metadata.is_dir() {
+            FileType::Directory
+        } else {
+            FileType::File
+        };
+
+        let lfs_state = match file_type {
+            FileType::File => {
+                let mut f = fs::File::open(file_path.clone())?;
+                // 19 bytes gets "version https://git", which is convincing
+                // enough to determine if a file is an LFS stub
+                let mut buffer = [0; 19];
+
+                f.read_exact(&mut buffer).ok();
+                if buffer.eq(b"version https://git") {
+                    LocalFileLFSState::Stub
+                } else {
+                    LocalFileLFSState::Local
+                }
+            }
+            FileType::Directory => LocalFileLFSState::None,
+        };
+
+        let size = match lfs_state {
+            LocalFileLFSState::Stub => {
+                let reader = BufReader::new(fs::File::open(file_path)?);
+
+                // get 3rd line, split by space, get 2nd element, parse to u64
+                let size: u64 = reader

```

Comment:
Not sure if Tauri gracefully handles the panics that might come from the unwraps here, but chaining `and_thens` might make it slightly easier to read. maybe something like:

`
reader
            .lines()
            .nth(2)
            .and_then(|line| line.ok())
            .and_then(|line| line.split(' ').nth(1))
            .and_then(|size_str| size_str.parse().ok())
            .ok_or_else(|| Err(create_tauri_error(res).await))?
`